src/Uno.UI/UI/Xaml/FrameworkElement.cs

```diff
@@ -752,13 +760,77 @@ internal override bool IsViewHit()
 
 		protected virtual bool GoToElementStateCore(string stateName, bool useTransitions) => false;
 
-		public event EventHandler<object> LayoutUpdated;
+		#region LayoutUpdated
 
-		internal virtual void OnLayoutUpdated()
+		private EventHandler<object> _layoutUpdated;
+
+		public event EventHandler<object> LayoutUpdated
 		{
-			LayoutUpdated?.Invoke(this, new RoutedEventArgs(this));
+			// This is the same as what the compiler generates for field-like events, except that we want to plug extra code (notifying EventManager).
+			add
+			{
+#if UNO_HAS_ENHANCED_LIFECYCLE
+				var isFirstSubscriber = _layoutUpdated is null;
+#endif
+
+				EventHandler<object> eventHandler = _layoutUpdated;
+				while (true)
+				{
+					EventHandler<object> eventHandler2 = eventHandler;
+					EventHandler<object> value2 = (EventHandler<object>)Delegate.Combine(eventHandler2, value);
+					eventHandler = Interlocked.CompareExchange(ref _layoutUpdated, value2, eventHandler2);
+					if ((object)eventHandler == eventHandler2)
+					{
+						break;
+					}
+				}
```

Comment:
You are never re-evaluating the `_layoutUpdated` in case of conflict, so you will never exit the loop.

BTW Is this from WinUI (which would explain the bug)? Otherwise the common names for such scenario are `original` and `updated`, suffixing with number is really not helpful.

```suggestion
				while (true)
				{
					var original = _layoutUpdated;
					var updated = (EventHandler<object>)Delegate.Combine(original, value);
					if (Interlocked.CompareExchange(ref _layoutUpdated, updated, original) == original)
					{
						break;
					}
				}
```

Reply:
@dr1rrb This is just the decompiled code when you have a regular field-like event in C#. https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgYRIG8CT6zSBTANyZgBcSBRNzgCQgwEAGyYAnADwB7YACsmAYw4A+EgBkIAT2kBXDgFUADgggcmCANwEAvkA=

Reply:
The whole idea behind the way the compiler generates this is thread-safety. I'm not sure if subscribing/unsubscribing to LayoutUpdated requires the UI thread in WinUI, in which case we can simplify this code. But for now, I preferred to keep it matching with what the compiler is lowering to.

Reply:
Oh indeed on line 781 the original `eventHandler` is being updated ... that why names matter! If it's from de-compiled *-**generated** code (so not from winUI) we need to give real names to variables.

And if it's only about replicates what's done by compiler, you can simply convert your private field `_layoutUpdated` to an `event` and do the `+=` and `-=` on it.

Reply:
I'm plugging extra logic so can't use a regular event. Will rename variables

Reply:
But you can still use a private event:, eg: 

``` csharp
		private event EventHandler<object> _layoutUpdated;

		public event EventHandler<object> LayoutUpdated
		{
			add
			{
#if UNO_HAS_ENHANCED_LIFECYCLE
				var isFirstSubscriber = _layoutUpdated is null;
#endif
				_layoutUpdated += value;
#if UNO_HAS_ENHANCED_LIFECYCLE
				if (isFirstSubscriber)
				{
					Uno.UI.Extensions.DependencyObjectExtensions.GetContext(this).EventManager.AddLayoutUpdatedEventHandler(this);
				}
#endif
			}
```

Reply:
Ah, yes. Indeed. Will do it that way to make the code cleaner.